crates/sui-core/src/signature_mpc/sign.rs

```diff
@@ -52,61 +53,92 @@ impl SignRound {
-
+        // let mut v = decryption_shares.clone();
+        // if party_id == 1 || party_id == 2 {
+        //     v[0] = (
+        //         PaillierModulusSizedNumber::from_u16(200),
+        //         PaillierModulusSizedNumber::from_u16(200),
+        //     );
+        // }
         Ok((
             SignRound::FirstRound {
                 signature_threshold_decryption_round_parties
             },
-            decryption_shares
+            decryption_shares.clone()
         ))
     }
 
-    pub(crate) fn complete_round(
-        &mut self,
-        state: SignState
-    ) -> Result<SignRoundCompletion> {
+    /// Tries to decrypt the signatures and return them.
+    /// In case one or more of the signatures is invalid, it will return a [`SignRoundCompletion::StartIAFlow`] to launch the IA flow.
+    pub(crate) fn complete_round(&mut self, state: SignState) -> Result<SignRoundCompletion> {
         let round = mem::take(self);
         match round {
-            SignRound::FirstRound { signature_threshold_decryption_round_parties } => {
-                let signatures_s = decrypt_signature_decentralized_party_sign(state.messages.unwrap(), state.tiresias_public_parameters.clone(), state.decryption_shares.clone(), state.public_nonce_encrypted_partial_signature_and_proofs.clone().unwrap(), signature_threshold_decryption_round_parties)?;
-
-                Ok(SignRoundCompletion::Output(signatures_s))            }
-            _ => Ok(SignRoundCompletion::None)
+            SignRound::FirstRound {
+                signature_threshold_decryption_round_parties,
+            } => {
+                let decrypt_result = decrypt_signature_decentralized_party_sign(
+                    state.messages.clone().unwrap(),
+                    state.tiresias_public_parameters.clone(),
+                    state.decryption_shares.clone(),
+                    state
+                        .public_nonce_encrypted_partial_signature_and_proofs
+                        .clone()
+                        .unwrap(),
+                    signature_threshold_decryption_round_parties,
+                );
+                match decrypt_result {
+                    Ok(signatures) => {
+                        Ok(SignRoundCompletion::SignatureOutput(signatures))
+                    }
+                    Err(decryption_error) => {
+                        Ok(SignRoundCompletion::StartIAFlow(
+                            decryption_error.failed_messages_indices,
+                            decryption_error.decrypters,
+                        ))
+                    }
+                }
+            }
+            _ => Ok(SignRoundCompletion::None),
         }
-
-
     }
 }
 
-
 pub(crate) enum SignRoundCompletion {
-    Output(Vec<Vec<u8>>),
+    SignatureOutput(Vec<Vec<u8>>),
+    StartIAFlow(Vec<usize>, Vec<PartyID>),
```

Comment:
I don't think it's good idea to use acronyms here - I would change it to InitiateIdentifiableAbort